kernel/src/fs/configfs/tsm.rs

```diff
@@ -0,0 +1,206 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{boxed::Box, string::ToString, sync::Arc};
+use core::fmt::Debug;
+
+use aster_systree::{
+    inherit_sys_branch_node, inherit_sys_leaf_node, BranchNodeFields, Error, NormalNodeFields,
+    Result, SysAttrSetBuilder, SysObj, SysPerms, SysStr,
+};
+use aster_util::printer::VmPrinter;
+use inherit_methods_macro::inherit_methods;
+use ostd::{
+    mm::{FallibleVmRead, FallibleVmWrite, VmReader, VmWriter},
+    sync::RwLock,
+};
+use spin::Once;
+
+use crate::device::tdxguest::tdx_get_quote;
+
+const TSM_INBLOB_MAX: usize = 64;
+
+#[derive(Debug)]
+struct Tsm {
+    fields: BranchNodeFields<ReportSystem, Self>,
+}
+
+impl Tsm {
+    fn new() -> Arc<Self> {
+        let name = SysStr::from("tsm");
+        let builder = SysAttrSetBuilder::new();
+
+        let attrs = builder.build().expect("Failed to build attribute set");
+
+        Arc::new_cyclic(|weak_self| {
+            let fields = BranchNodeFields::new(name, attrs, weak_self.clone());
+            fields
+                .add_child(ReportSystem::new())
+                .expect("Failed to add report directory");
+
+            Tsm { fields }
+        })
+    }
+}
+
+inherit_sys_branch_node!(Tsm, fields, {
+    fn perms(&self) -> SysPerms {
+        SysPerms::DEFAULT_RW_PERMS
+    }
+});
+
+#[derive(Debug)]
+struct ReportSystem {
+    fields: BranchNodeFields<Report, Self>,
+}
+
+#[inherit_methods(from = "self.fields")]
+impl ReportSystem {
+    fn new() -> Arc<Self> {
+        let name = SysStr::from("report");
+        let builder = SysAttrSetBuilder::new();
+
+        let attrs = builder.build().expect("Failed to build attribute set");
+
+        Arc::new_cyclic(|weak_self| ReportSystem {
+            fields: BranchNodeFields::new(name, attrs, weak_self.clone()),
+        })
+    }
+
+    fn add_child(&self, new_child: Arc<Report>) -> Result<()>;
+}
+
+inherit_sys_branch_node!(ReportSystem, fields, {
+    fn perms(&self) -> SysPerms {
+        SysPerms::DEFAULT_RW_PERMS
+    }
+
+    fn create_child(&self, name: &str) -> Result<Arc<dyn SysObj>> {
+        let report = Report::new(SysStr::from(name.to_string()));
+        self.add_child(report.clone())?;
+        Ok(report)
+    }
+});
+
+#[derive(Debug)]
+pub struct TsmProvider {
+    provider: SysStr,
+    inblob: [u8; TSM_INBLOB_MAX],
+    inblob_len: usize,
+    outblob: Option<Box<[u8]>>,
+    generation: u32,
+}
+
+impl TsmProvider {
+    pub fn new() -> Self {
+        TsmProvider {
+            provider: SysStr::from("tdx_guest"),
+            inblob: [0u8; TSM_INBLOB_MAX],
+            inblob_len: 0,
+            outblob: None,
+            generation: 1,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Report {
```

Comment:
Simply calling this struct `Report` may give people the wrong impression that it is the attestation report itself, which is not true. this Rename this struct to `ReportNode`.